Devnet transition updates

jagerman opened the pull request:
Updates from a deep hole of passing through the SENT code.

(Merged the changes from #1678 as well to get this conflict free)

Summary of changes (more details in individual commit messages):

- Get rid of std::strings of hex usage for BLS types; we now keep everything in crypto::bls_{pubkey,signature} and crypto::eth_address.  There was *some* of this before, but this completes the transition.

- Make `initCurve()` for herumi code thread-safe and initialized wherever it might be needed.

- Rework BLS private key generation and initialization to be done at the same time as other secret/public keys.  This also makes *all* keys types able to be stored as `0x`-prefixed hex (previously we required binary for the `key` and `key_ed25519` files).

- Add a C++ version of `crypto::keccak` that makes computing hashes of random types joined together much easier; this avoids the need to pre-concatenate the signing message in most cases.

- guts.h got some major improvements to make parsing binary packed values considerably easier, most notably split_guts_into and split_hex_guts_into.

- Moved and renamed `hex_to_type` and `hex_to_type` into guts.h, to be alongside the very similar view_guts and make_from_guts functions.

- Add utilities (`common/bigint.h`) for dealing with large encoded integers (such as the big-endian u256's that are incoming from ethereum)

- Make bls_utils and BLSSigner more usable with crypto::bls_whatever types.

- Split HF20 into HF20 + HF21: HF20 is for the transition period for us to get everything we need for seeding the reward contracts, while HF21 is where things are now running on ethereum.  (NB: existing code should prefer `feature::ETH_BLS` for code that should take effect under ETH; this also renamed hf20 with a suffix to be sure that no existing code is misusing it for something that should now be under HF21).

- Consistently use `bls_pubkey` instead of `bls_key`, particularly in RPC requests, but also in function names.  The `key` term is ambiguous, but worse, we have places in the code where "something_key" implies private key, so having it be public key here is a potential source of confusion.

- Change liquidation behaviour to use a network-configuration `ETH_EXIT_BUFFER` delay rather than the node's recommission credit.

- Rewrite uptime proofs to:
    - use more efficient (and easier to follow and validate) `oxenc::bt_dict_consumer` and `_producer` to parse and create proofs.
    - add BLS pubkey & signature into proofs during HF20 (and only HF20; we don't need them in HF21 because the bls keys will become part of the registration info).
    - fix a bug where new fields couldn't be added to proofs without breaking signature verification.  (Unfortunately this means we can't add the BLS key/signature until *after* HF20 because it would break under HF19 to add the extra data).
    - Disable differing primary and Ed25519 pubkeys in proofs starting at HF21.  Like BLS keys, the primary node will get updated to the ed pubkey in registrations and local keys at the transition block.
    - Drop the deprecated primary SN pubkey signature at HF21 (since we'll always have the proper Ed25519 signature and pubkeys will no longer be allowed to differ).

- Simplify log parsing by using the new guts.h splitting code.

- Simplify `std::optional<TransactionStateChangeVariant>` into `std::variant<std::monostate, ...>` (which is designed to be the same thing, but with non-values being just a state instead of an extra type layer).

- Fixed a bunch of RPC endpoints unconditionally setting binary values as always hex rather than going through `.response_hex` to set hex only when making non-binary requests.  `x.response_hex["foo"] = some_binary_value;` is equivalent to `x.response["foo"] = tools::hex_guts(some_binary_value)` when under json, but will be more efficient binary if making a bt-encoded binary request.

- Fix a shadowing bug in the RPC get_accrued_earnings that was always setting specifically requested address balances to 0.

- Remove defunct `ethereum_address_notification` tx type.  It was an idea from several months ago that doesn't have a place in our current transition/smart contract.

- Remove old (pre-HF19) non-bt-encoded uptime proof code.

- OMQ BLS request handlers:
    - move from cryptonote_core into bls_aggregator, to keep them alongside the code where we *make* the requests to those same endpoints.
    - Use bt-encoding of responses instead of ZMQ frame parts as this makes parsing a little more convenient, but more importantly makes the responses extensible in the future if we want to add fields of extra info.
    - Remove the BLS pubkey from responses: we already know the pubkey of who we're talking to, and should verify the signature against that rather than what they send back.
    - Add more key types that we need for proper verifications to the `service_node_address` struct.
    - Drop the signed message from BLS signing requests: like the pubkey, we already know exactly what they should be signing and can verify against it.
    - Combined nearly-identical exit and liquidation request handling
    - Add missing signature verification to exit & liquidation requests
    - Some code deduplication/mild refactoring.
    - Left some FIXMEs about needing to make this code asynchronous.
    - renamed `processNodes` to `nodesRequest` to better reflect what it does.
    - Fixed a bug in reward handling that was reusing the same local loop variable (by reference) for all incoming responses.
    - Removed the "add our own value to the result" code because it isn't necessarily correct: we could easily be making this call from a non-service node, or from an unregistered service node.  (Additionally oxenmq is smart about not actually making a connection if we request something from ourselves and just handling it via an already-present inproc socket).
        - Removed the `exclude` parameter for avoiding ourself: after the above change it wasn't being used anymore.

- A small smattering of fixes, tweaks, C++20 niceties, and so on etc.

jagerman committed:
Make crypto::eth_address and ::bls_public_key hashable

eth_address is a bit odd in that its 20-byte size isn't a multiple of 8,
and thus doesn't have size_t alignment, which means on 64-bit platforms
we have to copy the first 8 bytes of its data into a size_t in
`raw_hasher` rather than just reinterpret_cast'ing (to avoid misaligned
pointer deferencing).

jagerman committed:
fix fmt path formatting

jagerman committed:
Do BLS key initialization with other keys; allow hex encoding

- Renames bls key from bls.key to key_bls to be consistent with the
  other SN keys.
- Reuse current file reading/writing code for other keys.
- Remove filesystem code from BLSSigner.
- Allow hex encoding of keys, with or without 0x prefix, and with or
  without \n suffix.

jagerman committed:
Make initCurve thread-safe and idempotent

jagerman committed:
Make keccak hash interface nicer

- Make the piecewise version take the hash size.  For some reason only
  the single-shot version took a hash size because who knows.
- Fix the header to have its own C++ ifdef for extern "C"
- Adds a nicer to use C++ wrapper around single and multi-part keccak
  that takes any number of spans.
- Simplify some ridiculous hashing in wallet2.  (This changes the hashed
  value, but that's fine because this is only ever used ephemerally, and
  already was different across endian/architectures).

jagerman committed:
Ignore uninitialized warning in cryptonote_core ctor

`m_blockchain_storage` is indeed uninitialized when used here during
construction, but we don't actually touch it until after it has been
initialized.

This points at real initialization problems in the code base, but not
something we're going to fix right now (and not a bug in the current
implementation).

jagerman committed:
Utils enhancements

guts.h:
    - add `hex_guts` for a view_guts + oxenc::to_hex combo
    - add `span_guts` for getting a span
string_util.h:
    - add `_us` and `_usv` for unsigned char strings and string view
      literals.

Doy-lee committed:
Handle unused-variable and unused vars warning in RPC server

Doy-lee committed:
Remove extraneous return at end of void function

Doy-lee committed:
Revert "Revert "Remove oxen address workaround for addresses stored in DB""

This reverts commit da3dec4e2048bd97cd8844bec26a12740550ec26.

Doy-lee committed:
Fix various bugs to get rewards claim working

Rewards claim was mistakenly working due to leaving a work-around commit
in the git-tree and mistakenly testing the python script and assuming it
worked. After backing out those changes I identified that rewards claim
was still broken due to several bugs of which I have debugged and
incorporated into this PR.

- Ethereum addresses stored in the DB are stored with a 0x prefix. This
  was omitted as we were passing address strings directly through to the
  DB. I've made this type safe and convert at the very last moment to
  ensure we don't hit this bug. In future, it'd be better to store binary
  in the DB.

- The exclude parameter when querying nodes was broken. The code was
  double aggregating the node that was requesting a rewards claim. This
  meant the generated aggregate signature did not match the smart
  contract. This is fixed with the correct break in the for loop.
  Additionally, I've added a trace log to dump out the aggregate pubkey
  and how many keys we aggregated.

- I've added more error checking in general to try and catch problems earlier.

jagerman committed:
Move/replace type_to_hex and hex_to_type

These were very similar to view_guts, aside from hex conversion; this
unifies the interfaces into guts.h which is now responsible for both
type-to-hex and hex-to-type conversions alongside type-to-byte and
byte-to-hex conversions.

The new methods are also more robust, and can view (or accept) `unsigned
char` or `std::byte` strings as well.

- make_from_hex_guts<T>(hex) -- constructs a T from its hex content
- hex_guts(x) -- converts an `x` to hex; replaces `type_to_hex`
- load_from_hex_guts(hex, x) -- loads `hex` into `x`; throws on error
- try_load_from_hex_guts(hex, x) -- loads `hex` into `x`; returns true
  on success, false on error.

jagerman committed:
guts.h: split_guts_into and split_hex_into

split_guts_into lets you split an incoming chunk of data into multiple
trivially copyable types, which is very useful for unpacking incoming
packaged ethereum data into hashes and pubkeys and whatnot.

For example:

    struct A { int16_t x; };
    struct B { char y[5]; };
    using C = std::array<char, 6>;
    using D = std::array<uint16_t, 4>;

    auto val = "\x2a\x00hello world!\x01\x00\x02\x00\x03\x00\x04\x00omg"s;
    auto [a, b, c, d, e] = split_guts_into<A, B, skip<1>, C, D, std::string_view>(val);

yields:

    a.x == 42
    b == {'h', 'e', 'l', 'l', 'o'}
    c == {'w', 'o', 'r', 'l', 'd', '!'}
    d == {1, 2, 3, 4}  (or {256, 512, 768, 1024} on a big-endian architecture)
    e == "omg"sv

split_hex_into works the same way, but while also decoding input as hex rather than raw bytes.

jagerman committed:
common/bigint.h: utilities for large encoded integers

This makes it easier to deal with incoming big or little-endian integers
such as the 32-byte encoded values we get from ethereum RPC endpoints.

- encode_integer_be<N>(value) returns a big-endian, N-byte packed
  representation of value (commonly N is 32 when dealing with ethereum).
- encode_integer_be<N>(hi, lo) does the same for converting a 128-bit
  quantity (represented as a pair of uint64_ts).
- decode_integer_be(input) and decode_integer_be128 reverse the process,
  taking an array of bytes and returning a uint64_t (throwing if the
  value is too large to fit in a uint64_t), or pair of uint64_ts
  (throwing if the value is bigger than fits in a 128-bit integer).
- Versions are provided with `_le` instead of `_be` in the name to do
  the same but with little-endian encoding of the data.

jagerman committed:
Various fixes and improvements towards HF21

(This may be the longest commit message I've ever written.  It was
basically impossible to separate all of these out into individual
commits because they are heavily intertwined.)

- Convert hex strings to crypto:: types throughout the ethereum code.
  This adds/uses new types `crypto::eth_address`,
  `crypto::bls_public_key`, and `crypto::bls_signature` which are byte
  arrays under the hood, but give us type safe storage and various
  conversion convenience.  (A couple of these already existed, but
  weren't being used much).

- add a bunch of convenience methods to bls_utils for dealing with the
  crypto:: types, such as functions to swap between them and herumi bls
  objects, and a function to verify a signature without needing to touch
  the herumi code.

- rework BLSSigner:
    - various functions now have overloads to work with crypto:: types
    - the buildTag function gets renamed to buildTagHex, and a
      buildTagHash version that builds the tag into a crypto::hash.
    - buildTag* now take the network type instead of chainID +
      contractAddress and look up the chainID + contractAddress from the
      cryptonote config constants (via the network type).
    - change `crypto::bytes<32>` to `crypto::hash`; the `bytes<N>` type
      is meant to be a base implementation type for type-safe quantities
      (e.g. crypto::hash), not a directly used type.

- remove `get_bls_pubkeys`; we have better ways now to get that list.

- rename "bls_key" in various places to "bls_pubkey" to be consistent
  with member variable names for other pubkey types.  ("key" is
  ambiguous when you have private and public keys, but worse, in some
  places we are using "pubkey" and "key", the latter referring to the
  private key).

- Add distinction between HF20 (transition) and HF21 (ETH).

- Add BLS pubkey + signature into the HF20 proof (but only HF20; this
  doesn't continue into HF21).  Note that the signature here is a proof
  of possession over the SN pubkey + BLS pubkey, but is not the same
  proof of possession we use with the smart contract because
  transitioning nodes don't get created with a ETH address for the
  operator or contributor contract -- rather we have to manually seed
  them during TGE from what gets broadcast during the HF20 proof period.

- Added an ETH_EXIT_BUFFER network config parameter that defines how
  many blocks we wait before allowing penalized liquidation of a node
  (instead of regular exit).  Previously the logic was using the
  remaining recommission credit, but that has potential weird
  implications (and complicates explanation).  The buffer defaults to 7
  days or blocks on MAINNET and 1 hour of blocks on testnet/devnet.

- As of HF21, service node uptime proofs may no longer have different
  primary SN and Ed25519 public keys.  (Completing this transition by
  updating registrations and updating the local service node keys at the
  HF21 rollover is still a TODO item).

- The primary pubkey signature is no longer included in uptime proofs in
  HF21+, instead we just use the Ed25519 signature (which, combined with
  the previous point, is enough).

- Refactored/modernized uptime proof code to use
  bt_dict_producer/bt_dict_consumer to produce and consume proofs more
  efficiently than the older bt_dict object + serialization approach.

- Fix a bug in bt-encoded uptime proofs that didn't calculate the hash
  of the actual received data, but rather reserialized it from data it
  understood, which meant it was impossible to add new fields to a proof
  *before* a hardfork because the reconstitution wouldn't be the same as
  what the sender originally sent and hashed.  We know hash the incoming
  data as is, and store that for verification rather than needing to
  reserialize.

  This bug means we have to be careful to avoid including the BLS pubkey
  and and signature until after HF20; the intended behaviour of
  bt-encoded proofs was that new fields could be included without
  breaking proof verification, but this will be fixed for future
  hardforks.

- Add a TODO & failing assertion for the currently broken
  prepare_registration command under HF21+.

- Refactor log parsing using (new) split_hex_into to greatly simplify
  log parsing by not needing to have a ton of moving substrings.

- Eliminate the `RewardsLogEntry` class; the only thing it added over
  the `LogEntry` base type was a couple of methods that can be more
  simply used as free functions taking a LogEntry.

- replace `std::optional<TransactionStateChangeVariant>` (which is a
  `std::optional<std::variant<...>>`) with a
  `std::variant<std::monostate, ...>` which is meant to do the
  equivalent thing a bit more simply.

- Left TODOs for making RPC requests asynchronous.

- Fixed various RPC responses improperly setting binary values as hex.
  We want them as hex only for json requests, but want to keep them
  binary for binary (e.g. bt-encoded) requests.  Many of these just
  required a change to set the relevant binary values via the
  response_hex wrapper (to be converted to hex later in the response
  handler, if necessary).

jagerman committed:
Fix rpc get_accrued_earnings always returning 0

The inner `amount` here was shadowing the outer `amount` that it was
meant to be setting.

(There's another bug here that ETH addresses aren't accepted, to be
fixed in a later commit).

jagerman committed:
Tighten up random iterator requirements

jagerman committed:
fix C++ keccak interface

It wasn't working properly with all intended compatible input types.

jagerman committed:
Remove ethereum_address_notification transaction type

This was originally meant to allow OXEN -> SENT conversion transactions,
but we will now do that as part of the HF21 rollover with registered
addresses via the SN swap registration program finalized during the HF20
transition hardfork.

jagerman committed:
Remove old non-bt-encoded uptime proof code

This is all dead code; all network proofs have been only bt-encoded
proofs for a while now.

jagerman committed:
OMQ BLS reworking

- Move OMQ request handlers into bls_aggregator, so that the code for
  request handling & request making lives side-by-side.
- Change response methods to use a bt-encoded dict response so that we
  can more easily extend or change it in the future by adding new keys
  without having to worry about fixed position encoding.
- Don't include the BLS pubkey in the responses: the caller should
  already know the BLS pubkey of who they are talking to, and so passing
  it back is just redundant.  (If it doesn't match, we'll catch that
  immedately because signature verification of the response will fail).
- Don't include the signed message (or even the hash of the message) in
  the response: the requestor already knows that the message should be,
  and shouldn't be trust the remote to provide the value.  (As with the
  pubkeys above, if the remote signs the wrong thing the signature will
  fail).
- some mild refactoring to reduce code duplication
- leave FIXMEs behind about needing to make this code asynchronous

jagerman committed:
Merge doyle-harden-omq-endpoints2 + related updates

- BLSAggregator::processNodes (now renamed nodesRequest) was broken in
  that all requests referenced and set the results into the *same* local
  `request_result` variable for all responses.

- renamed `get_accrued_earnings_eth` and `get_accrued_earnings_oxen` to
  the same (overloaded) `get_accrued_earnings` name since they now have
  type-safe arguments.

- undo some of the code in bls_aggregator, particularly around the "add
  our own value to the result" because this method is callable on a
  non-service node, and is callable on an inactive service node, both of
  which would invalidate the signature.  (oxenmq also is perfectly happy
  to connect to itself by pubkey: it doesn't actually connect but uses
  an always-present internal inproc socket for handling such an internal
  request).

- Dropped oxenc::hex_to_type/type_to_hex; partly because of
  oxen-encoding PR comments, but also partly because much of that is
  superceded by new split hex to guts code.

- this also makes the `exclude` parameter unnecessary since we never
  exclude anything now, so dropped it.

jagerman committed:
Fix missing omq port in proof; catch Proof constructor exception

The required "q" parameter (oxend omq port) was missing.

Also noticed here as a result: currently the exception propagates back
to epee levin code; catch and log it and return false here instead.

jagerman committed:
Shorten l2 tracker sleep time

I'll revisit to avoid a dedicated thread at all when doing async
changes, but for now this shortening this reduces the devnet restart
time.

jagerman committed:
Require 0 storage server ports for DEVNET

DEVNET doesn't have a storage server, and so the check for 0 needs to be
outside the Proof constructor where we know the nettype.

jagerman committed:
Uncatch exceptions in Proof constructor

We were only catching to add a log warning (and then rethrowing), but
it's easier to not do that and have the parent catch and log as
appropriate.

(The `""sv` added here is to convince clang-format that formatting this
line by breaking the string "pke" into "pk" on one line and "e" on the
next is not actually a sensible way to do it.)

jagerman committed:
Fix backwards condition

jagerman committed:
Downgrade log spam

jagerman committed:
Allow last_decommission_reason_consensus_any to be omitted

The RPC server doesn't set these if they don't have a value.

jagerman committed:
Unbreak syncing, difficulty, etc.

`OXEN_USE_LOCAL_DEVNET_PARAMS` was *always* being defined for
compilation as this code wasn't checking the value of the cmake option.

jagerman committed:
Fix wrong height in log statement

This triggers when we are adding block x, so printing x - 1 in the
statement shows the wrong height.

jagerman committed:
Remove unused function

jagerman committed:
Fix eth_address being uninitialized causing junk in the batching db

jagerman committed:
Fix corrupted output in `print_pool` output

`tools::split` returns strings_views, but the dump result here wasn't
valid beyond the statement.

jagerman committed:
Fix get_transaction_pool endpoint

The missing declaration in the header meant that the overload for
GET_TRANSACTION_POOL was never called (but instead just the base class
GET_TRANSACTIONS was invoked), and so `request.memory_pool = true;`
didn't happen, and thus the endpoint would always return an empty list.

This fixes it, and also removes some commented out code and an unused
argument.

Doy-lee commented on src/common/formattable.h:
```diff
@@ -62,14 +60,105 @@ struct to_string_formatter : fmt::formatter<std::string_view> {
+// For example, for input span \0\0\0\0\0\0\0\x02 `fmt::format("0x{:z} 0x{:x} {}", val)` would
+// produced the string `"0x2 0x0000000000000002 0000000000000002"`.
+//
+// Some types extend this to override the default format -- for example, crypto::eth_address's
```
This footnote with special behaviour for `eth_address` seems a little bit tricky. It's a special case that's makes the API behave less predictable. TBF I believe this is only used at the fringe of the codebase, e.g. when displaying to the user, maybe at JSON serialisation due to ETH APIs requiring strict `0x` prefixes but it's in those cases where I'd want to be very explicit about adding a '0x' prefix or not.

Doy-lee commented on src/crypto/base.h:
```diff
@@ -56,18 +55,20 @@ auto operator<=>(const T& left, const T& right) {
-    size_t operator()(const T& val) const { return *reinterpret_cast<const size_t*>(val.data()); }
+    size_t operator()(const T& val) const {
+        if constexpr (alignof(T) >= sizeof(size_t))
+            return *reinterpret_cast<const size_t*>(val.data());
```
Smells like UB? Original backing memory (`unsigned char` essentially) was not declared as `size_t`. Just defaulting to `memcpy` seems fine.

jagerman committed:
Fix get_transactions rpc staked_amount value

- `staked_amount` was never set because the `tx.extra` had been cleared
  earlier.

Also fixes `last_failed_height` being copy-and-pasted wrong in the rpc
docs.

jagerman committed:
Fix broken oxend command-line commands

The RPC return value of these wasn't checked properly: `invoke` returns
a json object, not a bool.

jagerman committed:
Fix broken BLS fork check

Even if this comparison was correct, it's misleading because what
matters here is when ETH_BLS happens, not when the batching rewards
height is done.

jagerman committed:
Bump ethyl for header mismatch avoidance

jagerman committed:
Catch & log uptime proof errors

jagerman committed:
Make periodic_task exception safe with error logging

If an exception is thrown the task does not get counted as completed and
will be tried again the next time it is invoked.

This also simplifies the random delay to be a single value because
having two values for the delay pointlessly adds a degree of freedom:
the delay distribution `a + U[b, c]` is equivalent to `(a+b) + U[0,
c-b]`, and so you can just give the base (a+b) and max random offset
(c-b) to achieve the exact same distribution in two parameters instead
of three.

jagerman committed:
epee: fix broken exception handling

Various bits of epee's tcp p2p code have worse-than-nothing exception
safety: they `try { }` (via a disgusting macro of course because this is
epee) but also use manual mutex lock/unlock calls that they don't
properly unlock on the exception path.  Thus an exception in various
call paths permanently deadlock the P2P server.  Yay epee!

jagerman committed:
Fix compress/uncompressed mismatch

Our crypto::bls_signature is in uncompressed form (to match everything
eth contract side), but this was making a compressed form which won't
work.

jagerman committed:
Fix BLS pubkey & PoP parsing

These had mismatched send/receive keys.

Also use the feature constants instead of raw hf version checks.

jagerman committed:
Remove hash rate periodic check

This is pointless with pulse.  (Pulse can, in theory, fail, but it's
rare enough that catching it through this misleading warning isn't very
useful).

jagerman committed:
Fix failing crypto formatting tests

Recent commits removed the `<>` from the default format of crypto types.

jagerman committed:
Fix BLS signature loading, and add BLS unit tests

We didn't have proper signature deserialization: we need to do the exact
same thing to deserialize a signature that we have to do for pubkey, but
on G2 instead of G1.

This implements it via a generic implementation that combines the pubkey
and signature serialization code.

jagerman committed:
Remove debugging
